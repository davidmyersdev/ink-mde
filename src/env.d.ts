declare module '*.css' {
  const value: string

  export default value
}

declare module '*.css?inline' {
  const value: string

  export default value
}
