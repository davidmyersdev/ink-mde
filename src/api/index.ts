export * from './destroy'
export * from './focus'
export * from './format'
export * from './get_doc'
export * from './insert'
export * from './load'
export * from './options'
export * from './reconfigure'
export * from './select'
export * from './selections'
export * from './update'
export * from './wrap'
