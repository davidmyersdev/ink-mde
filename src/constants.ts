export const HYDRATION_MARKER = 'data-ink-mde-ssr-hydration-marker'
export const HYDRATION_MARKER_SELECTOR = `[${HYDRATION_MARKER}]`

export const getHydrationMarkerProps = () => {
  if (import.meta.env.VITE_SSR) {
    return {
      [HYDRATION_MARKER]: true,
    }
  }

  return {}
}
