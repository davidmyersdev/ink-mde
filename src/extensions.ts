import { closeBrackets as autocompleteExtension } from '@codemirror/autocomplete'
import { Compartment, EditorState } from '@codemirror/state'
import { vim } from '@replit/codemirror-vim'
import { isAutoDark } from './ui/utils'
import { dark, light } from '/src/vendor/extensions/appearance'
import { attribution as attributionExtension } from '/src/vendor/extensions/attribution'
import { images as imagesExtension } from '/src/vendor/extensions/images'
import { spellcheck as spellcheckExtension } from '/src/vendor/extensions/spellcheck'
import * as InkValues from '/types/values'

import type * as Ink from '/types/ink'
import type InkInternal from '/types/internal'

export const buildVendor = (extension: InkInternal.OptionExtension<Ink.Values.Extensions>, options: Ink.OptionsResolved) => {
  const result = extension.resolver(options)

  return extension.compartment.of(result)
}

export const buildVendors = (state: InkInternal.StateResolved) => {
  return state.extensions.map((extension) => {
    return buildVendor(extension, state.options)
  })
}

export const buildVendorUpdate = (extension: InkInternal.OptionExtension<Ink.Values.Extensions>, options: Ink.OptionsResolved) => {
  const result = extension.resolver(options)

  return extension.compartment.reconfigure(result)
}

export const buildVendorUpdates = (state: InkInternal.StateResolved) => {
  return state.extensions.map((extension) => {
    return buildVendorUpdate(extension, state.options)
  })
}

export const create = <T extends Ink.Options.ExtensionNames>(name: T): InkInternal.OptionExtension<T> => {
  const compartment = new Compartment()
  const resolver = resolvers[name]

  return {
    compartment,
    name,
    resolver,
  }
}

export const createExtensions = () => {
  return [
    create(InkValues.Extensions.Appearance),
    create(InkValues.Extensions.Attribution),
    create(InkValues.Extensions.Autocomplete),
    create(InkValues.Extensions.Images),
    create(InkValues.Extensions.ReadOnly),
    create(InkValues.Extensions.Spellcheck),
    create(InkValues.Extensions.Vim),
  ]
}

export const resolvers: InkInternal.Vendor.ExtensionResolvers = {
  appearance({ interface: { appearance } }: Ink.OptionsResolved) {
    if (appearance === InkValues.Appearance.Dark)
      return dark()
    if (appearance === InkValues.Appearance.Light)
      return light()

    // Automatically determine the correct color scheme.
    return isAutoDark() ? dark() : light()
  },
  attribution(options: Ink.OptionsResolved) {
    return options.interface.attribution ? attributionExtension() : []
  },
  autocomplete(options: Ink.OptionsResolved) {
    return options.interface.autocomplete ? autocompleteExtension() : []
  },
  images(options: Ink.OptionsResolved) {
    return options.interface.images ? imagesExtension() : []
  },
  readonly(options: Ink.OptionsResolved) {
    return options.interface.readonly ? EditorState.readOnly.of(true) : EditorState.readOnly.of(false)
  },
  spellcheck(options: Ink.OptionsResolved) {
    return options.interface.spellcheck ? spellcheckExtension() : []
  },
  vim(options: Ink.OptionsResolved) {
    return options.vim ? vim() : []
  },
}
