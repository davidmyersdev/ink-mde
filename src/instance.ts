import { override } from './utils/deepmerge'
import { toCodeMirror, toInk } from '/src/adapters/selections'
import { buildVendorUpdates } from '/src/extensions'
import * as formatter from '/src/formatter'
import { makeState } from '/src/vendor/state'

import type * as Ink from '/types/ink'
import type InkInternal from '/types/internal'

export const destroy = ([state, _setState]: InkInternal.Store) => {
  const { editor } = state()

  editor.destroy()
}

export const doc = ([state, _setState]: InkInternal.Store) => {
  const { editor } = state()

  return editor.state.sliceDoc()
}

export const focus = ([state, _setState]: InkInternal.Store) => {
  const { editor } = state()

  if (!editor.hasFocus)
    editor.focus()
}

export const format = ([state, setState]: InkInternal.Store, type: `${Ink.Values.Markup}`, selection?: Ink.Editor.Selection) => {
  return formatter.format([state, setState], type, selection)
}

export const insert = ([state, setState]: InkInternal.Store, text: string, selection?: Ink.Editor.Selection, updateSelection = false) => {
  const { editor } = state()

  let start = selection?.start
  let end = selection?.end || selection?.start

  if (typeof start === 'undefined') {
    const current = selections([state, setState]).pop() as Ink.Editor.Selection

    start = current.start
    end = current.end
  }

  const updates = { changes: { from: start, to: end, insert: text } }

  if (updateSelection) {
    const anchor = start === end ? start + text.length : start
    const head = start === end ? start + text.length : start + text.length

    Object.assign(updates, { selection: { anchor, head } })
  }

  editor.dispatch(
    editor.state.update(updates),
  )
}

export const load = ([state, setState]: InkInternal.Store, doc: string) => {
  setState(override(state(), { options: { doc } }))

  state().editor.setState(makeState(state()))
}

export const makeInstance = (store: InkInternal.Store): Ink.Instance => {
  return {
    destroy: destroy.bind(undefined, store),
    doc: doc.bind(undefined, store),
    focus: focus.bind(undefined, store),
    insert: insert.bind(undefined, store),
    load: load.bind(undefined, store),
    reconfigure: reconfigure.bind(undefined, store),
    select: select.bind(undefined, store),
    selections: selections.bind(undefined, store),
    update: update.bind(undefined, store),
    wrap: wrap.bind(undefined, store),
  }
}

export const reconfigure = ([state, setState]: InkInternal.Store, options: Ink.Options) => {
  const effects = buildVendorUpdates(setState(override(state(), { options })))

  state().editor.dispatch({
    effects,
  })
}

export const select = ([state, _setState]: InkInternal.Store, selections: Ink.Editor.Selection[]) => {
  const { editor } = state()

  editor.dispatch(
    editor.state.update({
      selection: toCodeMirror(selections),
    }),
  )
}

export const selections = ([state, _setState]: InkInternal.Store): Ink.Editor.Selection[] => {
  const { editor } = state()

  return toInk(editor.state.selection)
}

export const update = ([state, _setState]: InkInternal.Store, doc: string) => {
  const { editor } = state()

  editor.dispatch(
    editor.state.update({
      changes: {
        from: 0,
        to: editor.state.doc.length,
        insert: doc,
      },
    }),
  )
}

export const wrap = ([state, setState]: InkInternal.Store, { after, before, selection: userSelection }: Ink.Instance.WrapOptions) => {
  const { editor } = state()

  const selection = userSelection || selections([state, setState]).pop() || { start: 0, end: 0 }
  const text = editor.state.sliceDoc(selection.start, selection.end)

  insert([state, setState], `${before}${text}${after}`, selection)
  select([state, setState], [{ start: selection.start + before.length, end: selection.end + before.length }])
}
