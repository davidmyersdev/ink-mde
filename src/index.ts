import { makeInstance } from '/src/instance'
import { makeState } from '/src/state'

import type * as Ink from '/types/ink'

export * from '/types/values'
export * from '/src/vendor/extensions/extension'

export const defineOptions = (options: Ink.Options): Ink.Options => {
  return options
}

export const ink = (target: HTMLElement, options: Ink.Options = {}): Ink.Instance => {
  const ref = makeState({ target, options })

  return makeInstance(ref)
}

export default ink
