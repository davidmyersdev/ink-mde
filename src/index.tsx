import { render } from 'solid-js/web'
import { makeStore } from '/src/store'
import { makeInstance } from '/src/instance'
import { App } from '/src/ui/app'
import type * as Ink from '/types/ink'

export * from '/types/values'
export * from '/src/vendor/extensions/extension'

export const defineOptions = (options: Ink.Options): Ink.Options => {
  return options
}

export const ink = (target: HTMLElement, options: Ink.Options = {}): Ink.Instance => {
  const store = makeStore(options)

  render(() => <App store={store} />, target)

  return makeInstance(store)
}

export default ink
