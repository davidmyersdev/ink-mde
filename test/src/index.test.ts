import { beforeEach, describe, expect, it, vi } from 'vitest'
import { mockAll } from '/test/helpers/dom'
import { ink } from '/src/index'
import example from '/test/assets/example.md?raw'

describe('ink', () => {
  beforeEach(() => {
    vi.restoreAllMocks()
    mockAll()
  })

  it('mounts without errors', () => {
    const instance = ink(document.createElement('div'))

    expect(instance).toBeDefined()
  })

  it('matches the given options', () => {
    const instance = ink(document.createElement('div'), {
      doc: '# Hello',
    })

    expect(instance.getDoc()).toEqual('# Hello')
  })

  it('can be reconfigured', () => {
    const instance = ink(document.createElement('div'), {
      interface: {
        appearance: 'light',
      },
    })

    instance.reconfigure({ interface: { appearance: 'light' } })

    expect(instance.options().interface.appearance).toEqual('light')
  })

  it('matches the snapshot', () => {
    const target = document.createElement('div')

    // Todo: Use a comprehensive starter doc.
    ink(target, {
      doc: example,
      interface: {
        images: true,
        toolbar: true,
      },
    })

    expect(target).toMatchSnapshot()
  })
})
