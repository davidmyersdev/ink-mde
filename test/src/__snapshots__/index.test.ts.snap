// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ink > matches the snapshot 1`] = `
<div>
  <div
    class="ink ink-mde"
  >
    <!--#-->
    <style>
      .ink {
  --ink-internal-border-radius: var(--ink-border-radius, 0.25rem);
  --ink-internal-color: var(--ink-color, currentColor);
  --ink-internal-flex-direction: var(--ink-flex-direction, column);
  --ink-internal-font-family: var(--ink-font-family, inherit);
  --ink-internal-block-background-color: var(--ink-block-background-color, #f5f5f5);
  --ink-internal-block-background-color-on-hover: var(--ink-block-background-color-on-hover, #e0e0e0);
  --ink-internal-block-max-height: var(--ink-block-max-height, 20rem);
  --ink-internal-block-padding: var(--ink-block-padding, 0.5rem);
  --ink-internal-code-background-color: var(--ink-code-background-color, var(--ink-internal-block-background-color));
  --ink-internal-code-color: var(--ink-code-color, inherit);
  --ink-internal-code-font-family: var(--ink-code-font-family, 'Monaco', Courier, monospace);
  --ink-internal-editor-font-size: var(--ink-editor-font-size, 1em);
  --ink-internal-editor-line-height: var(--ink-editor-line-height, 2em);
  --ink-internal-editor-padding: var(--ink-editor-padding, 0.5rem);
  --ink-internal-inline-padding: var(--ink-inline-padding, 0.125rem);
  --ink-internal-modal-position: var(--ink-modal-position, fixed);
  --ink-internal-syntax-atom-color: var(--ink-syntax-atom-color, #d19a66);
  --ink-internal-syntax-comment-color: var(--ink-syntax-comment-color, #abb2bf);
  --ink-internal-syntax-comment-font-style: var(--ink-syntax-comment-font-style, italic);
  --ink-internal-syntax-emphasis-color: var(--ink-syntax-emphasis-color, inherit);
  --ink-internal-syntax-emphasis-font-style: var(--ink-syntax-emphasis-font-style, italic);
  --ink-internal-syntax-hashtag-background-color: var(--ink-syntax-hashtag-background-color, #eee);
  --ink-internal-syntax-hashtag-color: var(--ink-syntax-hashtag-color, inherit);
  --ink-internal-syntax-heading-color: var(--ink-syntax-heading-color, inherit);
  --ink-internal-syntax-heading-font-weight: var(--ink-syntax-heading-font-weight, 600);
  --ink-internal-syntax-heading1-color: var(--ink-syntax-heading1-color, var(--ink-internal-syntax-heading-color, inherit));
  --ink-internal-syntax-heading1-font-size: var(--ink-syntax-heading1-font-size, 1.6em);
  --ink-internal-syntax-heading1-font-weight: var(--ink-syntax-heading1-font-weight, 600);
  --ink-internal-syntax-heading2-color: var(--ink-syntax-heading2-color, var(--ink-internal-syntax-heading-color, inherit));
  --ink-internal-syntax-heading2-font-size: var(--ink-syntax-heading2-font-size, 1.5em);
  --ink-internal-syntax-heading2-font-weight: var(--ink-syntax-heading2-font-weight, 600);
  --ink-internal-syntax-heading3-color: var(--ink-syntax-heading3-color, var(--ink-internal-syntax-heading-color, inherit));
  --ink-internal-syntax-heading3-font-size: var(--ink-syntax-heading3-font-size, 1.4em);
  --ink-internal-syntax-heading3-font-weight: var(--ink-syntax-heading3-font-weight, 600);
  --ink-internal-syntax-heading4-color: var(--ink-syntax-heading4-color, var(--ink-internal-syntax-heading-color, inherit));
  --ink-internal-syntax-heading4-font-size: var(--ink-syntax-heading4-font-size, 1.3em);
  --ink-internal-syntax-heading4-font-weight: var(--ink-syntax-heading4-font-weight, 600);
  --ink-internal-syntax-heading5-color: var(--ink-syntax-heading5-color, var(--ink-internal-syntax-heading-color, inherit));
  --ink-internal-syntax-heading5-font-size: var(--ink-syntax-heading5-font-size, 1.2em);
  --ink-internal-syntax-heading5-font-weight: var(--ink-syntax-heading5-font-weight, 600);
  --ink-internal-syntax-heading6-color: var(--ink-syntax-heading6-color, var(--ink-internal-syntax-heading-color, inherit));
  --ink-internal-syntax-heading6-font-size: var(--ink-syntax-heading6-font-size, 1.1em);
  --ink-internal-syntax-heading6-font-weight: var(--ink-syntax-heading6-font-weight, 600);
  --ink-internal-syntax-highlight-background-color: var(--ink-syntax-highlight-background-color, #555555);
  --ink-internal-syntax-keyword-color: var(--ink-syntax-keyword-color, #c678dd);
  --ink-internal-syntax-link-color: var(--ink-syntax-link-color, inherit);
  --ink-internal-syntax-meta-color: var(--ink-syntax-meta-color, #abb2bf);
  --ink-internal-syntax-monospace-color: var(--ink-syntax-monospace-color, var(--ink-internal-code-color));
  --ink-internal-syntax-monospace-font-family: var(--ink-syntax-monospace-font-family, var(--ink-internal-code-font-family));
  --ink-internal-syntax-name-color: var(--ink-syntax-name-color, #d19a66);
  --ink-internal-syntax-name-label-color: var(--ink-syntax-name-label-color, #abb2bf);
  --ink-internal-syntax-name-property-color: var(--ink-syntax-name-property-color, #96c0d8);
  --ink-internal-syntax-name-property-definition-color: var(--ink-syntax-name-property-definition-color, #e06c75);
  --ink-internal-syntax-name-variable-color: var(--ink-syntax-name-variable-color, #e06c75);
  --ink-internal-syntax-name-variable-definition-color: var(--ink-syntax-name-variable-definition-color, #e5c07b);
  --ink-internal-syntax-name-variable-local-color: var(--ink-syntax-name-variable-local-color, #d19a66);
  --ink-internal-syntax-name-variable-special-color: var(--ink-syntax-name-variable-special-color, inherit);
  --ink-internal-syntax-number-color: var(--ink-syntax-number-color, #d19a66);
  --ink-internal-syntax-operator-color: var(--ink-syntax-operator-color, #96c0d8);
  --ink-internal-syntax-processing-instruction-color: var(--ink-syntax-processing-instruction-color, #bbbbbb);
  --ink-internal-syntax-punctuation-color: var(--ink-syntax-punctuation-color, #abb2bf);
  --ink-internal-syntax-strikethrough-color: var(--ink-syntax-strikethrough-color, inherit);
  --ink-internal-syntax-strikethrough-text-decoration: var(--ink-syntax-strikethrough-text-decoration, line-through);
  --ink-internal-syntax-string-color: var(--ink-syntax-string-color, #98c379);
  --ink-internal-syntax-string-special-color: var(--ink-syntax-string-special-color, inherit);
  --ink-internal-syntax-strong-color: var(--ink-syntax-strong-color, inherit);
  --ink-internal-syntax-strong-font-weight: var(--ink-syntax-strong-font-weight, 600);
  --ink-internal-syntax-url-color: var(--ink-syntax-url-color, #666666);
}

    </style>
    <!--/-->
    <!--#-->
    
    <!--/-->
    <!--#-->
    <div
      class="ink-mde-toolbar"
    >
      <style />
      <div
        class="ink-mde-container"
      >
        <div
          class="ink-mde-toolbar-group"
        >
          <!--#-->
          <button
            class="ink-button"
            type="button"
          >
            <svg
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-miterlimit="5"
              viewBox="0 0 20 20"
            >
              <path
                d="M6 4V10M6 16V10M6 10H14M14 10V4M14 10V16"
              />
            </svg>
          </button>
          <!--/-->
          <!--#-->
          <button
            class="ink-button"
            type="button"
          >
            <svg
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-miterlimit="5"
              stroke-width="1.5"
              viewBox="0 0 20 20"
            >
              <path
                d="M6.5 10H10.5C12.1569 10 13.5 11.3431 13.5 13C13.5 14.6569 12.1569 16 10.5 16H6.5V4H9.5C11.1569 4 12.5 5.34315 12.5 7C12.5 8.65686 11.1569 10 9.5 10"
              />
            </svg>
          </button>
          <!--/-->
          <!--#-->
          <button
            class="ink-button"
            type="button"
          >
            <svg
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-miterlimit="5"
              viewBox="0 0 20 20"
            >
              <path
                d="M11 4L9 16M13 4H9M7 16H11"
              />
            </svg>
          </button>
          <!--/-->
        </div>
        <div
          class="ink-mde-toolbar-group"
        >
          <!--#-->
          <button
            class="ink-button"
            type="button"
          >
            <svg
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-miterlimit="5"
              viewBox="0 0 20 20"
            >
              <path
                d="M2.00257 16H17.9955M2.00055 4H18M7 10H18.0659M2 8.5V11.4999C2.4 11.5 2.5 11.5 2.5 11.5V11V10.5M4 8.5V11.4999H4.5V11V10.5"
              />
            </svg>
          </button>
          <!--/-->
          <!--#-->
          <button
            class="ink-button"
            type="button"
          >
            <svg
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-miterlimit="5"
              viewBox="0 0 20 20"
            >
              <path
                d="M13 4L7 16"
              />
              <path
                d="M5 7L2 10L5 13"
              />
              <path
                d="M15 7L18 10L15 13"
              />
            </svg>
          </button>
          <!--/-->
          <!--#-->
          <button
            class="ink-button"
            type="button"
          >
            <svg
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-miterlimit="5"
              viewBox="0 0 20 20"
            >
              <path
                d="M7 4L8 6"
              />
            </svg>
          </button>
          <!--/-->
        </div>
        <div
          class="ink-mde-toolbar-group"
        >
          <!--#-->
          <button
            class="ink-button"
            type="button"
          >
            <svg
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-miterlimit="5"
              viewBox="0 0 20 20"
            >
              <path
                d="M7 16H17.8294"
              />
              <path
                d="M2 16H4"
              />
              <path
                d="M7 10H17.8294"
              />
              <path
                d="M2 10H4"
              />
              <path
                d="M7 4H17.8294"
              />
              <path
                d="M2 4H4"
              />
            </svg>
          </button>
          <!--/-->
          <!--#-->
          <button
            class="ink-button"
            type="button"
          >
            <svg
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-miterlimit="5"
              viewBox="0 0 20 20"
            >
              <path
                d="M7 16H18"
              />
              <path
                d="M2 17.0242C2.48314 17.7569 3.94052 17.6154 3.99486 16.7919C4.05315 15.9169 3.1975 16.0044 2.99496 16.0044M2.0023 14.9758C2.48544 14.2431 3.94282 14.3846 3.99716 15.2081C4.05545 16.0831 3.1998 16.0002 2.99726 16.0002"
              />
              <path
                d="M7 10H18"
              />
              <path
                d="M2.00501 11.5H4M2.00193 8.97562C2.48449 8.24319 3.9401 8.38467 3.99437 9.20777C4.05259 10.0825 2.04342 10.5788 2 11.4996"
              />
              <path
                d="M7 4H18"
              />
              <path
                d="M2 5.5H4M2.99713 5.49952V2.5L2.215 2.93501"
              />
            </svg>
          </button>
          <!--/-->
          <!--#-->
          <button
            class="ink-button"
            type="button"
          >
            <svg
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-miterlimit="5"
              viewBox="0 0 20 20"
            >
              <path
                d="M7 16H17.8294"
              />
              <path
                d="M5 15L3 17L2 16"
              />
              <path
                d="M7 10H17.8294"
              />
              <path
                d="M5 9L3 11L2 10"
              />
              <path
                d="M7 4H17.8294"
              />
              <path
                d="M5 3L3 5L2 4"
              />
            </svg>
          </button>
          <!--/-->
        </div>
        <div
          class="ink-mde-toolbar-group"
        >
          <!--#-->
          <button
            class="ink-button"
            type="button"
          >
            <svg
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-miterlimit="5"
              viewBox="0 0 20 20"
            >
              <path
                d="M9.12127 10.881C10.02 11.78 11.5237 11.7349 12.4771 10.7813L15.2546 8.00302C16.2079 7.04937 16.253 5.54521 15.3542 4.6462C14.4555 3.74719 12.9512 3.79174 11.9979 4.74539L10.3437 6.40007M10.8787 9.11903C9.97997 8.22002 8.47626 8.26509 7.52288 9.21874L4.74545 11.997C3.79208 12.9506 3.74701 14.4548 4.64577 15.3538C5.54452 16.2528 7.04876 16.2083 8.00213 15.2546L9.65633 13.5999"
              />
            </svg>
          </button>
          <!--/-->
          <!--#-->
          <button
            class="ink-button"
            type="button"
          >
            <svg
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-miterlimit="5"
              viewBox="0 0 20 20"
            >
              <rect
                height="12"
                rx="1"
                width="16"
                x="2"
                y="4"
              />
              <path
                d="M7.42659 7.67597L13.7751 13.8831M2.00208 12.9778L7.42844 7.67175"
              />
              <path
                d="M11.9119 12.0599L14.484 9.54443L17.9973 12.9785"
              />
              <path
                d="M10.9989 7.95832C11.551 7.95832 11.9986 7.52072 11.9986 6.98092C11.9986 6.44113 11.551 6.00354 10.9989 6.00354C10.4468 6.00354 9.99921 6.44113 9.99921 6.98092C9.99921 7.52072 10.4468 7.95832 10.9989 7.95832Z"
              />
            </svg>
          </button>
          <!--/-->
          <!--#-->
          
          <!--/-->
        </div>
      </div>
    </div>
    <!--/-->
    <div
      class="ink-mde-editor"
    >
      <div
        class="cm-editor ͼ1 ͼ2 ͼ4b ink-mde-container"
      >
        <div
          aria-live="polite"
          style="position: fixed; top: -10000px;"
        />
        <div
          class="cm-scroller"
          tabindex="-1"
        >
          <div
            aria-multiline="true"
            autocapitalize="off"
            autocorrect="off"
            class="cm-content cm-lineWrapping ink-mde-editor-content"
            contenteditable="true"
            data-language="markdown"
            role="textbox"
            spellcheck="false"
            style="tab-size: 4"
            translate="no"
          >
            <div
              class="cm-line"
            >
              <span
                class="ͼ3w ͼ4a"
              >
                #
              </span>
              <span
                class="ͼ3w"
              >
                 ink-mde
              </span>
            </div>
            <div
              class="cm-line"
            >
              <br />
            </div>
            <div
              class="cm-line"
            >
              A beautiful, modern, customizable Markdown editor powered by CodeMirror 6 and TypeScript.
            </div>
            <div
              class="cm-line"
            >
              <br />
            </div>
            <div
              class="cm-line"
            >
              Images can be previewed in-place. 
              <span
                class="ͼ46 ͼ4a"
              >
                ![
              </span>
              <span
                class="ͼ46"
              >
                kitten
              </span>
              <span
                class="ͼ46 ͼ4a"
              >
                ]
              </span>
              <span
                class="ͼ46 ͼ4a"
              >
                (
              </span>
              <span
                class="ͼ46 ͼ47"
              >
                https://placekitten.com/2000/200
              </span>
              <span
                class="ͼ46 ͼ4a"
              >
                )
              </span>
            </div>
            <div
              class="cm-line"
            >
              <br />
            </div>
            <div
              class="cm-line"
            >
              Inline code (e.g. 
              <span
                class="cm-code cm-code-open"
                data-enable-grammarly="false"
                data-gramm="false"
                data-grammarly-skip="true"
                spellcheck="false"
              >
                <span
                  class="ͼ4a"
                >
                  \`
                </span>
              </span>
              <span
                class="cm-code"
                data-enable-grammarly="false"
                data-gramm="false"
                data-grammarly-skip="true"
                spellcheck="false"
              >
                <span
                  class="ͼ3m"
                >
                  const a = 10
                </span>
              </span>
              <span
                class="cm-code cm-code-close"
                data-enable-grammarly="false"
                data-gramm="false"
                data-grammarly-skip="true"
                spellcheck="false"
              >
                <span
                  class="ͼ4a"
                >
                  \`
                </span>
              </span>
              ) and code blocks are highlighted in-place.
            </div>
            <div
              class="cm-line"
            >
              <br />
            </div>
            <div
              class="cm-codeblock cm-codeblock-open cm-line"
              data-enable-grammarly="false"
              data-gramm="false"
              data-grammarly-skip="true"
              spellcheck="false"
            >
              <span
                class="ͼ4a"
              >
                \`\`\`
              </span>
              <span
                class="ͼ3o"
              >
                ts
              </span>
            </div>
            <div
              class="cm-codeblock cm-line"
              data-enable-grammarly="false"
              data-gramm="false"
              data-grammarly-skip="true"
              spellcheck="false"
            >
              // Write code with syntax highlighting.
            </div>
            <div
              class="cm-codeblock cm-line"
              data-enable-grammarly="false"
              data-gramm="false"
              data-grammarly-skip="true"
              spellcheck="false"
            >
              const add = (a: number, b: number): number =&gt; {
            </div>
            <div
              class="cm-codeblock cm-line"
              data-enable-grammarly="false"
              data-gramm="false"
              data-grammarly-skip="true"
              spellcheck="false"
            >
                return a + b
            </div>
            <div
              class="cm-codeblock cm-line"
              data-enable-grammarly="false"
              data-gramm="false"
              data-grammarly-skip="true"
              spellcheck="false"
            >
              }
            </div>
            <div
              class="cm-codeblock cm-codeblock-close cm-line"
              data-enable-grammarly="false"
              data-gramm="false"
              data-grammarly-skip="true"
              spellcheck="false"
            >
              <span
                class="ͼ4a"
              >
                \`\`\`
              </span>
            </div>
            <div
              class="cm-line"
            >
              <br />
            </div>
            <div
              class="cm-line"
            >
              Stay 
              <span
                class="ͼ3i ͼ4a"
              >
                *
              </span>
              <span
                class="ͼ3i"
              >
                focused
              </span>
              <span
                class="ͼ3i ͼ4a"
              >
                *
              </span>
               on what you write instead of breaking concentration to check formatting.
            </div>
            <div
              class="cm-line"
            >
              <br />
            </div>
            <div
              class="cm-line"
            >
              <span
                class="ͼ4a"
              >
                -
              </span>
               
              <span
                class="ͼ46 ͼ4a"
              >
                [
              </span>
              <span
                class="ͼ46"
              >
                voracious
              </span>
              <span
                class="ͼ46 ͼ4a"
              >
                ]
              </span>
              <span
                class="ͼ46 ͼ4a"
              >
                (
              </span>
              <span
                class="ͼ46 ͼ47"
              >
                https://github.com/voracious
              </span>
              <span
                class="ͼ46 ͼ4a"
              >
                )
              </span>
            </div>
            <div
              class="cm-line"
            >
              <span
                class="ͼ4a"
              >
                -
              </span>
               
              <span
                class="ͼ46 ͼ4a"
              >
                [
              </span>
              <span
                class="ͼ46"
              >
                voracious/ink-mde
              </span>
              <span
                class="ͼ46 ͼ4a"
              >
                ]
              </span>
              <span
                class="ͼ46 ͼ4a"
              >
                (
              </span>
              <span
                class="ͼ46 ͼ47"
              >
                https://github.com/voracious/ink-mde
              </span>
              <span
                class="ͼ46 ͼ4a"
              >
                )
              </span>
            </div>
            <div
              class="cm-line"
            >
              <span
                class="ͼ4a"
              >
                -
              </span>
               
              <span
                class="ͼ46 ͼ4a"
              >
                [
              </span>
              <span
                class="ͼ46"
              >
                voracious/octo
              </span>
              <span
                class="ͼ46 ͼ4a"
              >
                ]
              </span>
              <span
                class="ͼ46 ͼ4a"
              >
                (
              </span>
              <span
                class="ͼ46 ͼ47"
              >
                https://github.com/voracious/octo
              </span>
              <span
                class="ͼ46 ͼ4a"
              >
                )
              </span>
            </div>
            <div
              class="cm-line"
            >
              <br />
            </div>
          </div>
        </div>
      </div>
    </div>
    <!--#-->
    <div
      class="ink-mde-details"
    >
      <div
        class="ink-mde-container"
      >
        <div
          class="ink-mde-details-content"
        >
          <!--#-->
          
          <!--/-->
          <!--#-->
          
          <!--/-->
          <!--#-->
          <div
            class="ink-mde-attribution"
          >
            <span>
               powered by 
              <a
                class="ink-mde-attribution-link"
                href="https://github.com/voracious/ink-mde"
                rel="noopener noreferrer"
                target="_blank"
              >
                ink-mde
              </a>
            </span>
          </div>
          <!--/-->
        </div>
      </div>
    </div>
    <!--/-->
  </div>
</div>
`;
